Client/Frontend/Brave Today/Composer/FeedDataSource+RSS.swift

```diff
@@ -0,0 +1,212 @@
+import Fuzi
+import Shared
+
+struct RSSFeedLocation: Hashable {
+    var title: String?
+    var url: URL
+    
+    var id: String {
+        url.absoluteString
+    }
+}
+
+extension FeedDataSource {
+    
+    // MARK: - RSS Sources
+    
+    var rssFeedLocations: [RSSFeedLocation] {
+        RSSFeedSource.all().compactMap {
+            guard let url = URL(string: $0.feedUrl) else { return nil }
+            return RSSFeedLocation(title: $0.title, url: url)
+        }
+    }
+    
+    /// Add a users custom RSS feed to the list of sources
+    ///
+    /// - returns: `true` if the feed is successfully added, `false` if it already exists
+    @discardableResult
+    func addRSSFeedLocation(_ location: RSSFeedLocation) -> Bool {
+        let feedUrl = location.url.absoluteString
+        if RSSFeedSource.get(with: feedUrl) != nil {
+            return false
+        }
+        RSSFeedSource.insert(title: location.title,
+                             feedUrl: feedUrl)
+        setNeedsReloadCards()
+        return true
+    }
+    
+    /// Remove a users custom RSS feed to the list of sources
+    func removeRSSFeed(with url: URL) {
+        let feedUrl = url.absoluteString
+        if RSSFeedSource.get(with: feedUrl) == nil {
+            return
+        }
+        RSSFeedSource.delete(with: url.absoluteString)
+        setNeedsReloadCards()
+    }
+    
+    /// Whether or not an RSS feed is currently enabled
+    ///
+    /// - note: RSS Feeds are enabled by default since they are added by the user
+    func isRSSFeedEnabled(_ location: RSSFeedLocation) -> Bool {
+        FeedSourceOverride.get(fromId: location.id)?.enabled ?? true
+    }
+    
+    /// Toggle an RSS feed enabled state
+    func toggleRSSFeedEnabled(_ location: RSSFeedLocation, enabled: Bool) {
+        FeedSourceOverride.setEnabled(forId: location.id, enabled: enabled)
+        setNeedsReloadCards()
+    }
+}
+
+extension FeedItem.Content {
+    init?(from feedItem: JSONFeedItem, location: RSSFeedLocation) {
+        guard let publishTime = feedItem.datePublished,
+              let url = feedItem.url?.asURL,
+              let title = feedItem.title else {
+            return nil
+        }
+        var description = ""
+        var imageURL: URL?
+        if let image = feedItem.image {
+            imageURL = URL(string: image, relativeTo: location.url.domainURL)
+        }
+        if let text = feedItem.contentText {
+            description = text
+        }
+        if let html = feedItem.contentHtml {
+            let doc = try? HTMLDocument(string: html)
+            if imageURL == nil, let src = doc?.firstChild(xpath: "//img[@src]")?.attr("src") {
+                imageURL = URL(string: src, relativeTo: location.url.domainURL)
+            }
+            if description.isEmpty, let text = doc?.root?.childNodes(ofTypes: [.Text, .Element]).map({ node in
+                node.stringValue
+                    .trimmingCharacters(in: .whitespacesAndNewlines)
```

Comment:
`text nodes` can have html content as well, may be `escapeHTML` here?

Reply:
I tested what `stringValue` returns here and it automatically decodes HTML entities (i.e. `&amp;` became `&`), so I don't think we need further calls